[clang] When checking for covariant return types, make sure the pointers or references are to *classes*

bricknerb opened the pull request:
https://eel.is/c++draft/class.virtual#8.1

This prevents overriding methods with non class return types that have less cv-qualification.

Fixes: #111742

bricknerb committed:
[clang] When checking for covariant return types, make sure the pointers or references are to *classes*.

https://eel.is/c++draft/class.virtual#8.1

This prevents overriding methods with non class return types that have less cv-qualification.

bricknerb committed:
Fix indentation.

bricknerb committed:
Merge branch 'main' into covariant

Sirraide reviewed (commented):
Implementation seems fine, but this is still missing a release note (in `clang/docs/ReleaseNotes.rst`, probably somewhere in the ‘Potentially Breaking Changes’ section).

CC @AaronBallman @cor3ntin Do we want to make this a warning that defaults to an error instead? Imo it’s probably fine as-is because I don’t imagine too many people would write code that abuses us not diagnosing this...

bricknerb committed:
[clang] Add the breaking change to more correctly check covariance when return type doesn't point to a class to release notes.

bricknerb committed:
[clang] Fix code indentation in release notes for fixing how we handle pointers to non classes when calculating covariance.

cor3ntin commented:
>  CC @AaronBallman @cor3ntin Do we want to make this a warning that defaults to an error instead? Imo it’s probably fine as-is because I don’t imagine too many people would write code that abuses us not diagnosing this...

It looks fine to me as is.

Code that would do something like `static_cast<BaseReturnType>(derived->f())` (which I think is perfectly reasonable)  would explode, despite the standard guarantees it should work.

If people scream we have plenty of time to reconsider making it a warning

cor3ntin commented on clang/test/SemaCXX/virtual-override.cpp:
```diff
@@ -289,3 +289,13 @@ namespace PR8168 {
   };
 }
+
+namespace T13 {
+  struct A {
+    virtual const int *f() const; // expected-note{{overridden virtual function is here}}
+  };
+
+  struct B : A {
+    int *f() const override; // expected-error{{virtual function 'f' has a different return type ('int *') than the function it overrides (which has return type 'const int *')}}
+  };
+}
```
can we have additional tests with various combination of `void*`, `int*` and some class type?

bricknerb replied:
Do you mean that base return type A (`void*`, `int*` or some class type) and derived returns type B != A (`void*`, `int*` or some class type)?
Or do you mean they both return the same type but it's qualified differently?

For different types we have T1 and T2.
For different qualification of class types, we have T6.
Let me know what use case you think is uncovered.

ilya-biryukov replied:
Given that the code has a different treatment of types which are pointers and references, having a variation of `T2` with non-class types seems valuable. (we could probably add it to `T2` directly).

Same for qualification in `T6`, given that the rules are special for the class types.

bricknerb replied:
Added tests in T2 and T6, and removed T13.
Please let me know if this is not what you meant.

bricknerb committed:
[clang] Unify the new covariant tests, remove T13 and add the new tests into T2 and T6.
In T2, we check that return types that are pointers to different non-class types fail.
In T6, we check that unlike class return types, non-class return types do not allow to change qualifiers at all (add or remove).

cor3ntin reviewed (approved):
LGTM

bricknerb committed:
Merge branch 'main' into covariant

ilya-biryukov reviewed (approved):
LGTM

bricknerb committed:
Merge branch 'main' into covariant
